AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/MutationEvent+Extensions.swift

```diff
@@ -30,11 +35,25 @@ extension MutationEvent {
                     return
                 }
 
-                if existingEvent.version == nil {
-                    Amplify.log.verbose("""
-                        Replacing existing mutation event having nil version with version from mutation response
-                            \(mutationSync.syncMetadata.version)
-                        """)
+                // return if version of the pending mutation event is not nil and
+                // is >= version contained in the response
+                if existingEvent.version != nil && existingEvent.version! >= mutationSync.syncMetadata.version {
+                    completion(.success(()))
+                    return
+                }
+
+                do {
+                    let responseModel = mutationSync.model.instance
+                    let requestModel = try mutationEvent.decodeModel()
+
+                    // check if the data sent in the request is the same as the response
+                    // if it is, update the pending mutation event version to the response version
+                    guard let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
+                          modelSchema.isEqual(responseModel, requestModel) else {
+                        completion(.success(()))
+                        return
+                    }
+
                     existingEvent.version = mutationSync.syncMetadata.version
```

Comment:
can this be factored out so it can be called like this? (and easier to test)

```swift
guard let reconciledEvent = reconcile(pendingMutationEvent: existingEvent, with: mutationEvent, responseMutationSync: mutationSync) else {
  complete(.success())
 return
}

 storageAdapter.save(reconciledEvent)
```
```swift
static func reconcile(pendingMutationEvent: MutationEvent, 
         with requestMutationEvent: MutationEvent, 
          responseMutationSync: MutationSync<AnyModel>) -> MutationEvent? {


 // return if version of the pending mutation event is not nil and
                // is >= version contained in the response
                if pendingMutationEvent.version != nil && pendingMutationEvent.version! >= mutationSync.syncMetadata.version {
                    return nil
                }

                do {
                    let responseModel = mutationSync.model.instance
                    let requestModel = try mutationEvent.decodeModel()

                    // check if the data sent in the request is the same as the response
                    // if it is, update the pending mutation event version to the response version
                    guard let modelSchema = ModelRegistry.modelSchema(from: mutationEvent.modelName),
                          modelSchema.isEqual(responseModel, requestModel) else {
                        return nil
                    }

                    pendingMutationEvent.version = mutationSync.syncMetadata.version
                 return pendingMutationEvent
          } catch {
               Amplify.log.verbose("Error decoding models: \(error)")
               return nil
          }
}


existing unit tests can be extended with one line:
```swift
// currently we unit test `reconcilePendingMutationEventsVersion` like so, followed by verifying the pending mutation queue
MutationEvent.reconcilePendingMutationEventsVersion(sent: request,
                                                             received: response,
                                                             storageAdapter: storageAdapter) 

// you can add an additional check like so:
let reconciledEvent = reconcile(pendingMutationEvent: head, with: mutationEvent, responseMutationSync: response)

// depending on the situation you would either
XCTAssertNil(reconciledEvent) // no version update
// or
XCTAssertEqual(reconciledEvent.version, response.version)
```